### ISSUE TITLE
[stmt.if] p2 Evaluations in the discarded substatement of constexpr if statement

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#268]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

Consider this example
````cpp
template<class T>
struct AlwaysInt{
   using type = int;
};
template<class T>
void fun(){
    if constexpr(std::is_same_v<T, char>){
        ;
   }else{
        typename AlwaysInt<T>::type i = 0;
        int c = 0;
        i++ + i; // #1
        c++ + c; // #2
   }
}
int main(){
    fun<char>(); 
}
````
In [stmt.if] p2, we says 
> if the condition is not value-dependent after its instantiation, the discarded substatement (if any) is not instantiated. 

Does `#2` cause the program to be undefined behavior if `T` is `char`? Is `#1` UB if it had the type that would be as if the substatement is instantiated? 

[intro.execution] p10 merely says
> If a side effect on a memory location ([intro.memory]) is unsequenced relative to either another side effect on the same memory location or a value computation using the value of any object in the same memory location, and they are not potentially concurrent ([intro.multithread]), the behavior is undefined.  

It seems to uniformly apply to all evaluations. It does not say whether it refers to these expressions that are executed/evaluated. Or,  even consider this example
````cpp
int main(){
   int i = 0;
   if(true){
      ;
   }else{
       i++ + i;  // control flow never goes through here
   }
}
````

**Comment by jensmaurer** [comment]
[intro.execution] p8 says

> Sequenced before is an asymmetric, transitive, pair-wise relation between evaluations executed by a single thread (6.9.2)

Note "executed". Obviously, the non-taken branch of an "if" is not executed, thus the evaluations contained therein aren't even part of the set on which "sequenced before" is defined.

Would it help to somehow integrate "executed" into the definition of "unsequenced"?

**Comment by xmh0511** [comment]
> Would it help to somehow integrate "executed" into the definition of "unsequenced"?

Yes, emphasizing the "execute" is helpful. Moreover, is the discarded substatement of `constexpr if` regulated by [stmt.if] p1? i.e. not be executed.

**Comment by frederick-vs-ja** [comment]
> > If a side effect on a memory location ([intro.memory]) is unsequenced relative to either another side effect on the same memory location or a value computation using the value of any object in the same memory location, and they are not potentially concurrent ([intro.multithread]), the behavior is undefined **if such two evaluations are executed**.

This seems still making unexecuted evaluations unsequenced (!?). I think sequence before, sequence after, indeterminately sequenced, and unsequenced relations are only meaningful for executed evaluations.

I guess it's better to say
> If <ins>**both _A_ and _B_ are executed,**</ins> _A_ is not sequenced before _B_<ins>**,**</ins> and _B_ is not sequenced before _A_, then _A_ and _B_ are _unsequenced_.

, which leaves unexecuted evaluations not falling into any of these relations.
